pkg/web/apps/apps_psql_repository.go

```diff
@@ -128,5 +130,154 @@ func (a *appsPostgreSQLRepository) GetAppByID(ID string) (*models.App, error) {
+
+	sqlStatement := `
+	SELECT v.id,v.version,v.app_id, v.disabled, v.disabled_message, v.num_of_app_launches, v.last_launched_at
+	FROM version as v
+	WHERE v.app_id = $1 AND v.version = $2;`
+
+	err := a.db.QueryRow(sqlStatement, appID, versionNumber).Scan(&version.ID, &version.Version, &version.AppID, &version.Disabled, &disabledMessage, &version.NumOfAppLaunches, &lastLaunchedAt)
+
+	version.DisabledMessage = disabledMessage.String
+	version.LastLaunchedAt = lastLaunchedAt.String
+
+	if err != nil {
+		log.Error(err)
+		if err == sql.ErrNoRows {
+			return nil, models.ErrNotFound
+		}
+		return nil, models.ErrInternalServerError
+	}
+
+	return &version, nil
+}
+
+// GetDeviceByDeviceIDAndAppID returns a device by its device ID and app ID
+func (a *appsPostgreSQLRepository) GetDeviceByDeviceIDAndAppID(deviceID string, appID string) (*models.Device, error) {
+	device := models.Device{}
+
+	sqlStatement := `
+		SELECT d.id, d.version_id, d.app_id, d.device_id, d.device_type, d.device_version
+		FROM device as d
+		WHERE d.device_id = $1 AND d.app_id = $2;`
+
+	if err := a.db.QueryRow(sqlStatement, deviceID, appID).
+		Scan(&device.ID, &device.VersionID, &device.AppID, &device.DeviceID, &device.DeviceType, &device.DeviceVersion); err != nil {
+
+		log.Error(err)
+		if err == sql.ErrNoRows {
+			return nil, models.ErrNotFound
+		}
+		return nil, models.ErrInternalServerError
+	}
+
+	return &device, nil
+}
+
+// GetDeviceByVersionAndAppID returns a device by its version number and app ID
+func (a *appsPostgreSQLRepository) GetDeviceByVersionAndAppID(version string, appID string) (*models.Device, error) {
+	device := models.Device{}
+
+	sqlStatement := `
+		SELECT d.id, d.version_id, d.app_id, d.device_id, d.device_type, d.device_version
+		FROM device as d
+		WHERE d.app_id = $1 AND d.device_version = $2;`
+
+	if err := a.db.QueryRow(sqlStatement, appID, version).
+		Scan(&device.ID, &device.VersionID, &device.AppID, &device.DeviceID, &device.DeviceType, &device.DeviceVersion); err != nil {
+
+		log.Error(err)
+		if err == sql.ErrNoRows {
+			return nil, models.ErrNotFound
+		}
+		return nil, models.ErrInternalServerError
+	}
+
+	return &device, nil
+}
+
+// GetAppByID retrieves an app by id from the database
+func (a *appsPostgreSQLRepository) GetAppByAppID(appID string) (*models.App, error) {
+	app := models.App{}
+
+	sqlStatment := `SELECT id,app_id,app_name FROM app WHERE app_id=$1 AND deleted_at IS NULL;`
+	err := a.db.QueryRow(sqlStatment, appID).Scan(&app.ID, &app.AppID, &app.AppName)
+
+	if err != nil {
+		log.Error(err)
+		if err == sql.ErrNoRows {
+			return nil, models.ErrNotFound
+		}
+		return nil, models.ErrInternalServerError
+	}
+
+	return &app, nil
+
+}
+
+// InsertVersionOrUpdateNumOfAppLaunches creates a new version row
+// or increments the num_of_app_launches counter if the version already exists
+func (a *appsPostgreSQLRepository) InsertVersionOrUpdateNumOfAppLaunches(version *models.Version) error {
```

Comment:
The name currently does not capture that last_launched_at is updated as well and it would become quite long if that is added as well. This comment is not very helpful :-D

Reply:
I will have a think for names. If you can think of any too that would be good - hard to find one :dagger:

Reply:
I was initially thinking of `UpsertVersion`, but this function is not that generic as we do not want to increment `num_of_app_launches` or set `last_launched_at` or any other update. It is only a small thing, so maybe something like `UpsertVersionWithAppLaunchesAndLastLaunched`.

Reply:
@davidffrench then IHMO it is the wrong layer. The logic over what should be done is part of the service and not the repo. Following my suggestion

apps_service.go
```go
func (a *appsService) InitClientApp(deviceInfo *models.Device) (*models.Version, error) {
	
	version, err := a.repository.GetVersionByAppIDAndVersion(deviceInfo) 

	// If any error other Not Found error occurred, return
	if err != nil && err != models.ErrNotFound {
		return nil, err
	}

	device, err := a.repository.GetDeviceByVersionIDandAppID(deviceInfo)

	// If is a new installation create the device and update the totals
	if err != nil && err != models.ErrNotFound {
		device, err := a.repository.CreateDevice(deviceInfo)
		if err != nil {
			return err
		}
		a.repository.IncrementVersionTotals(version, false) 
	}
	//If is not a new installation then just update the totals 
	a.repository.IncrementVersionTotals(version, false)	
	return version, nil
}

func (a *appsService) cleanVersionResponseForInitCall(version *models.Version) (*models.Version, error) {
	//TODO: Remove from the answer the totals before return this information should not be sent to the mobile device 
}

```

Repository
```go
func (a *appsPostgreSQLRepository) IncrementVersionTotals(versionInfo models.Version, isNewInstall bool) error {

	if isNewInstall {
		//TODO: It will update the TOTALS in the version table adding +1 for new install and launches
	}


	if !isNewInstall {
		//TODO: It will update the TOTALS in the version table adding +1 just for the launches
	}
	
	return nil
}
```

Reasons for this approach
* Make clear and less code required
* It has high cohesion 
* It is following the good practices and respecting the layers.
* if you add the totals in the service in the struct you will be more susceptible to face concurrency issues since between the time to get the object and the value in this layer until the moment of persisting the data it could, in fact, be already changed. 

I hope that it help

Reply:
Hi @camilamacedo86 .

I really like the structure of the code, specifically calling out hasVersion and hasDevice, I think it is easier to read.

Unfortunately there is no way around having the `/init` create the version, this is the only time we will get to know a new version automatically without manual input from a user.

The sample code in your comment does not cover all scenarios either, I have updated the flow chart to account for all flows of the init endpoint. The DB creates/updates could esaily be inserts as well.

![mss init flow](https://user-images.githubusercontent.com/1596014/53870228-b356f700-3ff1-11e9-9f93-c31aa349460b.png)

Reply:
@davidffrench  I checked the chart added here. I could not fund one scenario that is not covered by the suggestion. Note that:

- if has a version and has deviceXVersion then it is the launch scenario 
- if has a version but has not deviceXVersion then   it is the new install
- if has NOT a version and NOT a deviceXVersion it is new install and a new version

What situation is not covered?

Reply:
@camilamacedo86 I don't see the last scenario covered where the versionID needs to be updated for a device if the versionID does not match on the retrieved device.

Reply:
Hi @davidffrench yes it us true. Following the same with this last scenario. 

#### If has a version and has tracked a device using this version then it is the launch scenario 
**Actions**
* update the total launch in the version table

#### If has a version but has NOT tracked a device using this version then it is the new install
**Actions**
* Check if the device exist in the database with any other version if the SAME app
* If NOT exist then -> create the device 
* If exist then -> update the version number in the device 
* update the total launch and install in the version table
####  If has NOT the version then it is a new device install of a new version
**Actions**
* create the version 
* create the device 
* update the totals

c/c @craicoverflow

Reply:
I will close this one because it is too big and also few requirements changed already. 
I will open a new conversation. 
c/c @davidffrench and @craicoverflow